### ISSUE TITLE
Task GetAntlrJar fails with a ANTLR 4.9

### ISSUE THREAD

**Issue by NiccoMlt** [kaby76/Antlr4BuildTasks#9]
After updating my dependency from `Antlr4.Runtime.Standard` 4.8 to 4.9, I get this error.

>   Antlr4BuildTasks.targets(143, 3): [MSB4018] errore imprevisto dell'attività "GetAntlrJar".
System.Exception: Unhandled version of Antlr4.Runtime.Standard. Make sure you add <PackageReference Include="Antlr4.Runtime.Standard" Version="4.8" /> to your csproj file.
   in Antlr4.Build.Tasks.GetAntlrJar.Execute()
   in Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute()
   in Microsoft.Build.BackEnd.TaskBuilder.<ExecuteInstantiatedTask>d__26.MoveNext()

Is there some hard-coded version check?

**Comment by kaby76** [comment]
Yes, it has some hard-coded string tests. I'm making a change to fix this, output a better message, and roll forward with version 4.9 of the Antlr tool. I'll make a release (v8.8) in a couple hours after testing. Thank you for raising this bug.

**Comment by kaby76** [comment]
I released v8.8 of Antlr4BuildTasks (and the templates). The default is now 4.9 Antlr, but it will work with older versions of the runtime. Let me know if that fixes the problem.

**Comment by NiccoMlt** [comment]
It seems to work now, thank you!

**Comment by NiccoMlt** [comment]
Hi, I'm sorry for bothering you again.

For some reasons I had to go back to ANTLR 4.8, but Antlr4BuildTasks 8.8.0 (after the fix of this issue) gives me the following error:

>   Antlr4BuildTasks.targets(143, 3): [MSB4018] errore imprevisto dell'attività "GetAntlrJar".
System.Exception: Cannot download version 4.8.0 of the Antlr toolset. Please check the version and https://www.antlr.org/download/index.html for an available '-complete.jar' file version. Make sure the version number is exact, e.g., '4.9', not '4.9.0'.
   in Antlr4.Build.Tasks.GetAntlrJar.Execute()
   in Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute()
   in Microsoft.Build.BackEnd.TaskBuilder.<ExecuteInstantiatedTask>d__26.MoveNext()

I wrote here because it may be related, but if you want I can open you a separate issue

**Comment by NiccoMlt** [comment]
Also, now that I cleaned up the NuGet cache and the various output folders, even if I switch back to ANTLR 4.9, Antlr4buildtasks 8.8.0 fails at some point with the following error on Visual Studio:

> MSB4181	The "RunAntlrTool" task returned false but did not log an error.

which on JetBrains Rider is:

>   Antlr4BuildTasks.targets(168, 3): [MSB4181] l'attività "RunAntlrTool" ha restituito false, ma non è stato registrato alcun errore.

I think my bad that I didn't configure a solid Continuous Integration and didn't spot the problem early.

**Comment by kaby76** [comment]
The first thing to do here is to do a sanity check. Please clone https://github.com/kaby76/testantlr then do a "dotnet build; dotnet run" in the directory containing the .csproj. It should build and run the expression grammar example.

**Comment by NiccoMlt** [comment]
The test kinda works:

> ❯ dotnet build
> Microsoft (R) Build Engine versione 16.8.0+126527ff1 per .NET
> Copyright (C) Microsoft Corporation. Tutti i diritti sono riservati.
> 
>   Individuazione dei progetti da ripristinare...
> C:\Program Files\dotnet\sdk\5.0.101\NuGet.targets(131,5): warning : No Authorization header detected [C:\Users\nicco\RiderProjects\testantlr\testantlr.csproj]
> C:\Program Files\dotnet\sdk\5.0.101\NuGet.targets(131,5): warning : No Authorization header detected [C:\Users\nicco\RiderProjects\testantlr\testantlr.csproj]
> C:\Program Files\dotnet\sdk\5.0.101\NuGet.targets(131,5): warning : No Authorization header detected [C:\Users\nicco\RiderProjects\testantlr\testantlr.csproj]
> C:\Program Files\dotnet\sdk\5.0.101\NuGet.targets(131,5): warning : No Authorization header detected [C:\Users\nicco\RiderProjects\testantlr\testantlr.csproj]
>   Tutti i progetti sono aggiornati per il ripristino.
> C:\Users\nicco\RiderProjects\testantlr\testantlr.csproj : warning Undefined: No Authorization header detected
> C:\Users\nicco\RiderProjects\testantlr\testantlr.csproj : warning Undefined: No Authorization header detected
> C:\Users\nicco\RiderProjects\testantlr\testantlr.csproj : warning Undefined: No Authorization header detected
> C:\Users\nicco\RiderProjects\testantlr\testantlr.csproj : warning Undefined: No Authorization header detected
>   testantlr -> C:\Users\nicco\RiderProjects\testantlr\bin\Debug\netcoreapp3.1\testantlr.dll
> 
> Compilazione completata.
> 
> C:\Program Files\dotnet\sdk\5.0.101\NuGet.targets(131,5): warning : No Authorization header detected [C:\Users\nicco\RiderProjects\testantlr\testantlr.csproj]
> C:\Program Files\dotnet\sdk\5.0.101\NuGet.targets(131,5): warning : No Authorization header detected [C:\Users\nicco\RiderProjects\testantlr\testantlr.csproj]
> C:\Program Files\dotnet\sdk\5.0.101\NuGet.targets(131,5): warning : No Authorization header detected [C:\Users\nicco\RiderProjects\testantlr\testantlr.csproj]
> C:\Program Files\dotnet\sdk\5.0.101\NuGet.targets(131,5): warning : No Authorization header detected [C:\Users\nicco\RiderProjects\testantlr\testantlr.csproj]
> C:\Users\nicco\RiderProjects\testantlr\testantlr.csproj : warning Undefined: No Authorization header detected
> C:\Users\nicco\RiderProjects\testantlr\testantlr.csproj : warning Undefined: No Authorization header detected
> C:\Users\nicco\RiderProjects\testantlr\testantlr.csproj : warning Undefined: No Authorization header detected
> C:\Users\nicco\RiderProjects\testantlr\testantlr.csproj : warning Undefined: No Authorization header detected
>     Avvisi: 8
>     Errori: 0
> 
> Tempo trascorso 00:00:02.26
> 
> ❯ dotnet run
> C:\Program Files\dotnet\sdk\5.0.101\NuGet.targets(131,5): warning : No Authorization header detected [C:\Users\nicco\RiderProjects\testantlr\testantlr.csproj]
> C:\Program Files\dotnet\sdk\5.0.101\NuGet.targets(131,5): warning : No Authorization header detected [C:\Users\nicco\RiderProjects\testantlr\testantlr.csproj]
> C:\Program Files\dotnet\sdk\5.0.101\NuGet.targets(131,5): warning : No Authorization header detected [C:\Users\nicco\RiderProjects\testantlr\testantlr.csproj]
> C:\Program Files\dotnet\sdk\5.0.101\NuGet.targets(131,5): warning : No Authorization header detected [C:\Users\nicco\RiderProjects\testantlr\testantlr.csproj]
> C:\Users\nicco\RiderProjects\testantlr\testantlr.csproj : warning Undefined: No Authorization header detected
> C:\Users\nicco\RiderProjects\testantlr\testantlr.csproj : warning Undefined: No Authorization header detected
> C:\Users\nicco\RiderProjects\testantlr\testantlr.csproj : warning Undefined: No Authorization header detected
> C:\Users\nicco\RiderProjects\testantlr\testantlr.csproj : warning Undefined: No Authorization header detected
> 1 + 2 + 3
> parse completed.
> 1 2 + 3
> line 1:2 mismatched input '2' expecting {<EOF>, ';'}
> line 1:2 mismatched input '2' expecting {<EOF>, ';'}
> error in parse.
> 1 + +
> line 1:5 extraneous input '<EOF>' expecting {VARIABLE, SCIENTIFIC_NUMBER, '+', '-'}
> line 1:5 extraneous input '<EOF>' expecting {VARIABLE, SCIENTIFIC_NUMBER, '+', '-'}
> error in parse.

The grammar shows some warning but the code is generated correctly here.

**Comment by NiccoMlt** [comment]
My solution is composed by many projects, but only a few use ANTLR and your task;
maybe I could try unloading the other projects and report you back

**Comment by kaby76** [comment]
"warning : No Authorization header detected" is a problem with nuget.exe. Sounds like it's executing an old version of nuget.exe. But, I'm not sure what dotnet.exe is actually spawning for nuget.exe. One thing I noticed and complained about recently is this crappy NET SDK update that came over with Windows 10 Update. Try a "which dotnet.exe" and see what it's picking up. Then do a search for "dotnet.exe" from "c:/". You might be having two dotnet.exe. You should be executing the 64-bit version, not the x86 32-bit one.

**Comment by kaby76** [comment]
I do notice that this example I just checked in should actually not work because it references in the .csproj version "4.9.0" not "4.9" for Antlr. It still works, but it really should not because there is no "antlr-4.9.0-complete.jar" only "antlr-4.9-complete.jar". I'll fix that after we figure out what your build is doing.

**Comment by NiccoMlt** [comment]
> My solution is composed by many projects, but only a few use ANTLR and your task;
maybe I could try unloading the other projects and report you back

I tried manually cleaning up everything and unloading every non-antlr-related projects and it seems to work

> "warning : No Authorization header detected" is a problem with nuget.exe. Sounds like it's executing an old version of nuget.exe

Actually, I thought it was related to the auto-generated file not having a standard header with license/authorization infos and some analyzer was complaining. Remeber that .NET 5 bundles many of them.
In my project I disabled header-related style rule via the EditorConfig and the warning is not showing up.

> One thing I noticed and complained about recently is this crappy NET SDK update that came over with Windows 10 Update. Try a "which dotnet.exe" and see what it's picking up. Then do a search for "dotnet.exe" from "c:/". You might be having two dotnet.exe. You should be executing the 64-bit version, not the x86 32-bit one.

Yes, recently they added .NET SDK to Windows Updates, but I manually and deliberately updated my SDK to .NET 5; `which` doesn't work on Windows, but the equivalent confirms that I should only have a dotnet.exe:

> ❯ Get-Command dotnet.exe
> 
> CommandType     Name                                               Version    Source
> -----------     ----                                               -------    ------
> Application     dotnet.exe                                         5.0.120... C:\Program Files\dotnet\dotnet.exe

**Comment by NiccoMlt** [comment]
> I do notice that this example I just checked in should actually not work because it references in the .csproj version "4.9.0" not "4.9" for Antlr. It still works, but it really should not because there is no "antlr-4.9.0-complete.jar" only "antlr-4.9-complete.jar". I'll fix that after we figure out what your build is doing.

Where are fetching the jar from? I think pulling it from Maven should be the best solution.

**Comment by kaby76** [comment]
Better than Maven stores, I downloaded the antlr jar from the antlr website.

Doesn't `get-command` tell what is avaliable, not specifically which .exe you are executing? Check PATH for multiple  `dotnet/` directories.

**Comment by kaby76** [comment]
You're probably right--I should grab the Antlr jar from Maven Central Repository. But, I don't know the full URL to any of the downloads, antlr-4.9-complete.jar. Searching everywhere for the damn link--nothing is easy, not a single, absolute, simple "https://.... " URL in https://search.maven.org/artifact/org.antlr/antlr4/4.9/jar webpage.

**Comment by NiccoMlt** [comment]
> Doesn't get-command tell what is avaliable, not specifically which .exe you are executing? 

Actually, from [docs](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/get-command) it looks really similar to what `which` does on Linux, and in fact it is [the usually suggested PowerShell alternative](https://stackoverflow.com/a/65148/7907339).

But, the following is the `where` output run from `cmd` instead of PowerShell:

```
C:\Users\nicco>where dotnet.exe
C:\Program Files\dotnet\dotnet.exe
```

And checking in both system and user-specific PATH variables, I can see only one `dotnet` folder.

> You're probably right--I should grab the Antlr jar from Maven Central Repository. 

> I see: https://search.maven.org/remotecontent?filepath=org/antlr/antlr4/4.9/antlr4-4.9-complete.jar. I assume similarly for 4.8, 4.7, etc. I will change this for the next release of Antlr4BuildTasks.

Great! I think it's a very good improvement

**Comment by NiccoMlt** [comment]
> 
> 
>     My solution is composed by many projects, but only a few use ANTLR and your task;
>     maybe I could try unloading the other projects and report you back
> 
> I tried manually cleaning up everything and unloading every non-antlr-related projects and it seems to work

Currently, it works, so I think I cleaned up too much or too little and Visual Studio was running your tasks in a weird way.
Sorry for bothering you again.

I can close this issue, or keep it open if you think it relates with the jar-fetching problem.

**Comment by kaby76** [comment]
I'll close this but will update the code to (1) use the Maven Central Repository for the download, (2) fix up the code that tests between "4.9" vs "4.9.0" for the jar, (3) improve exception handling error messages.

**Comment by NiccoMlt** [comment]
Thanks again!
